src/compiler/visitorPublic.ts

```diff
@@ -378,929 +378,799 @@ namespace ts {
-                    nodeVisitor(node.questionToken, tokenVisitor, isQuestionToken),
-                    nodeVisitor(node.whenTrue, visitor, isExpression),
-                    nodeVisitor(node.colonToken, tokenVisitor, isColonToken),
-                    nodeVisitor(node.whenFalse, visitor, isExpression));
-
-            case SyntaxKind.TemplateExpression:
-                Debug.type<TemplateExpression>(node);
-                return factory.updateTemplateExpression(node,
-                    nodeVisitor(node.head, visitor, isTemplateHead),
-                    nodesVisitor(node.templateSpans, visitor, isTemplateSpan));
-
-            case SyntaxKind.YieldExpression:
-                Debug.type<YieldExpression>(node);
-                return factory.updateYieldExpression(node,
-                    nodeVisitor(node.asteriskToken, tokenVisitor, isAsteriskToken),
-                    nodeVisitor(node.expression, visitor, isExpression));
-
-            case SyntaxKind.SpreadElement:
-                Debug.type<SpreadElement>(node);
-                return factory.updateSpreadElement(node,
-                    nodeVisitor(node.expression, visitor, isExpression));
-
-            case SyntaxKind.ClassExpression:
-                Debug.type<ClassExpression>(node);
-                return factory.updateClassExpression(node,
-                    nodesVisitor(node.modifiers, visitor, isModifierLike),
-                    nodeVisitor(node.name, visitor, isIdentifier),
-                    nodesVisitor(node.typeParameters, visitor, isTypeParameterDeclaration),
-                    nodesVisitor(node.heritageClauses, visitor, isHeritageClause),
-                    nodesVisitor(node.members, visitor, isClassElement));
-
-            case SyntaxKind.ExpressionWithTypeArguments:
-                Debug.type<ExpressionWithTypeArguments>(node);
-                return factory.updateExpressionWithTypeArguments(node,
+    type VisitEachChildFunction<T extends Node> = (node: T, visitor: Visitor, context: TransformationContext, nodesVisitor: NodesVisitor, nodeVisitor: NodeVisitor, tokenVisitor: Visitor | undefined) => T;
+
+    // Constructs a type that correlates a `SyntaxKind` to either:
+    //   - `VisitEachChildFunction<T>`, if the `Node` for that kind has children
+    //   - `undefined`, if the `Node` for that kind has no children
+    //
+    // This looks something like:
+    //
+    //  {
+    //      [SyntaxKind.Unknown]?: undefined;
+    //      ...
+    //      [SyntaxKind.QualifiedName]: VisitEachChildFunction<QualifiedName>;
+    //  }
+    //
+    // This is then used as the expected type for `visitEachChildTable`.
+    type VisitEachChildTable =
+        & { [K in Exclude<SyntaxKind, HasChildren["kind"]>]?: undefined; } // nodes that don't have children are ignored
+        & { [TNode in HasChildren as TNode["kind"]]: VisitEachChildFunction<TNode> } // nodes that can have children must have methods

```

Comment:
Do you really need the first part of the intersection? Won't excess property checks catch anything extra?

Reply:
You need the first part of the intersection otherwise you can't index into `visitEachChildTable` with `kind` unless you've already checked `kind` is in `HasChildren["kind"]`.

Reply:
You need to assert on the use-site, just simplify here.

Reply:
Removing this would require the following change, which I believe is far less safe due to the `any` cast:

```ts
// was:
const fn = visitEachChildTable[node.kind] as VisitEachChildFunction<T> | undefined;

// suggested:
const fn = (visitEachChildTable as any)[node.kind] as VisitEachChildFunction<T> | undefined;
```

The type for `visitEachChildTable` as currently written accurately reflects the type:
- Any `SyntaxKind` that is also in `HasChildren` will have a `VisitEachChildFunction` value.
- Any other `SyntaxKind` will return `undefined`.

Reply:
Just write

`(visitEachChildTable as Record<SyntaxKind, VisitEachChildFunction | undefined>)[node.kind]`

or

```ts
visitEachChildTable[node.kind as keyof typeof visitEachChildTable]
```

Reply:
I'll have to use the first one then. The second one doesn't work.